File: frontend/listImage/src/domain/entity/listImage.ts
// frontend/listImage/src/domain/entity/listImage.ts

/**
 * ListImage
 * backend/internal/domain/listImage/entity.go の ListImage に対応。
 *
 * - 日付は ISO8601 文字列（例: "2025-01-10T00:00:00Z"）を想定
 * - updatedAt / updatedBy / deletedAt / deletedBy は省略可
 */
export interface ListImage {
  id: string;
  listId: string;
  url: string;
  fileName: string;
  size: number;
  displayOrder: number;

  createdAt: string;
  createdBy: string;
  updatedAt?: string | null;
  updatedBy?: string | null;
  deletedAt?: string | null;
  deletedBy?: string | null;
}

/**
 * 画像ファイルの検証結果（UI 向け）
 * backend/internal/domain/listImage/entity.go の ImageFileValidation に対応。
 */
export interface ImageFileValidation {
  isValid: boolean;
  errorMessage?: string;
}

/**
 * UI 向けエラーメッセージ（backend と整合）
 */
export const ERR_MSG_INVALID_FILE_TYPE = "無効なファイル形式です";
export const ERR_MSG_FILE_TOO_LARGE = "ファイルサイズが大きすぎます";
export const ERR_MSG_UPLOAD_FAILED = "画像のアップロードに失敗しました";

/**
 * NewImageFileValidation 相当
 */
export function newImageFileValidation(err?: Error | null): ImageFileValidation {
  if (!err) return { isValid: true };
  return { isValid: false, errorMessage: err.message };
}

/**
 * Policy
 * backend/internal/domain/listImage/entity.go の定数と整合させる。
 */
export const DEFAULT_MAX_IMAGE_SIZE_BYTES = 5 * 1024 * 1024; // 5MB

export const SUPPORTED_IMAGE_MIMES: ReadonlySet<string> = new Set([
  "image/jpeg",
  "image/jpg",
  "image/png",
  "image/webp",
]);

// Allowed file extensions for ListImage（空なら制限なし）
export const ALLOWED_EXTENSIONS: ReadonlySet<string> = new Set([
  ".png",
  ".jpg",
  ".jpeg",
  ".webp",
  ".gif",
]);

// 0 の場合は上限なし（Go 実装では 20MB 想定）
export const MAX_FILE_SIZE = 20 * 1024 * 1024; // 20MB

/**
 * 必須文字列チェック（service からも利用可能なユーティリティ）
 */
export function requireNonEmpty(name: string, v: string): Error | null {
  if (!v || v.trim() === "") {
    return new Error(`${name} is required`);
  }
  return null;
}

/**
 * data URL 形式 (data:<mime>;base64,...) の簡易検証
 * - backend の ValidateDataURL と整合する範囲でフロント用に実装
 * - payload は base64 デコード済み Uint8Array を返す（環境により利用可）
 */
export function validateDataUrl(
  data: string,
  maxBytes: number = DEFAULT_MAX_IMAGE_SIZE_BYTES,
  supported: ReadonlySet<string> = SUPPORTED_IMAGE_MIMES
): { mime: string; payload: Uint8Array } {
  if (!data.startsWith("data:")) {
    throw new Error("invalid data URL: missing 'data:' prefix");
  }

  const [meta, raw] = data.split(",", 2);
  if (!raw) {
    throw new Error("invalid data URL: missing payload");
  }
  if (!meta.includes(";base64")) {
    throw new Error("invalid data URL: not base64 encoded");
  }

  const mime = meta.substring("data:".length, meta.indexOf(";base64"));
  if (!mime) {
    throw new Error("invalid data URL: missing mime type");
  }
  if (!supported.has(mime)) {
    throw new Error(`unsupported content type: ${mime}`);
  }

  // base64 デコード（ブラウザ前提）
  let binary = "";
  try {
    binary = atob(raw);
  } catch (e) {
    throw new Error("invalid base64 payload");
  }

  const len = binary.length;
  if (len === 0) {
    throw new Error("empty image payload");
  }
  if (len > maxBytes) {
    throw new Error(`file too large: ${len} bytes (max ${maxBytes})`);
  }

  const bytes = new Uint8Array(len);
  for (let i = 0; i < len; i++) {
    bytes[i] = binary.charCodeAt(i);
  }

  return { mime, payload: bytes };
}

/**
 * URL の簡易バリデーション
 * backend の validateURL と整合する範囲。
 */
export function isValidUrl(u: string): boolean {
  if (!u || !u.trim()) return false;
  try {
    const parsed = new URL(u);
    return !!parsed.protocol && !!parsed.host;
  } catch {
    return false;
  }
}

/**
 * 拡張子チェック
 */
export function isAllowedExtension(fileName: string): boolean {
  if (!fileName) return false;
  if (ALLOWED_EXTENSIONS.size === 0) return true;
  const idx = fileName.lastIndexOf(".");
  if (idx < 0) return false;
  const ext = fileName.slice(idx).toLowerCase();
  return ALLOWED_EXTENSIONS.has(ext);
}

/**
 * ListImage のバリデーション
 * backend/internal/domain/listImage/entity.go の validate() と可能な範囲で整合。
 */
export function validateListImage(li: ListImage): boolean {
  // id / listId
  if (!li.id || !li.id.trim()) return false;
  if (!li.listId || !li.listId.trim()) return false;

  // url
  if (!isValidUrl(li.url)) return false;

  // fileName
  if (!li.fileName || !isAllowedExtension(li.fileName)) return false;

  // size
  if (typeof li.size !== "number" || li.size < 0) return false;
  if (MAX_FILE_SIZE > 0 && li.size > MAX_FILE_SIZE) return false;

  // displayOrder
  if (!Number.isInteger(li.displayOrder) || li.displayOrder < 0) return false;

  // createdAt / createdBy
  if (!li.createdAt || Number.isNaN(Date.parse(li.createdAt))) return false;
  if (!li.createdBy || !li.createdBy.trim()) return false;

  // updatedAt
  if (li.updatedAt != null && li.updatedAt !== "" &&
      Number.isNaN(Date.parse(li.updatedAt))) {
    return false;
  }

  // deletedAt
  if (li.deletedAt != null && li.deletedAt !== "" &&
      Number.isNaN(Date.parse(li.deletedAt))) {
    return false;
  }

  return true;
}

File: backend/internal/domain/list/list_image.go
// backend/internal/domain/list/list_image.go
package list

import (
	"encoding/base64"
	"errors"
	"fmt"
	"strings"
	"time"
)

// ListImage is an image record under:
//
//	/lists/{listId}/images/{imageId}
//
// Firebase Storage 移行後:
// - URL is Firebase Storage getDownloadURL()
// - ObjectPath is Firebase Storage object path
// - backend only persists and reads image metadata
// - backend does not upload/delete GCS objects
type ListImage struct {
	ID           string    `json:"id"`
	ListID       string    `json:"listId"`
	URL          string    `json:"url"`
	ObjectPath   string    `json:"objectPath"`
	FileName     string    `json:"fileName"`
	ContentType  string    `json:"contentType,omitempty"`
	Size         int64     `json:"size"`
	DisplayOrder int       `json:"displayOrder"`
	CreatedAt    time.Time `json:"createdAt"`
	CreatedBy    string    `json:"createdBy,omitempty"`
	UpdatedAt    time.Time `json:"updatedAt,omitempty"`
	UpdatedBy    string    `json:"updatedBy,omitempty"`
}

// ImageFileValidation - 画像ファイルのバリデーション結果
type ImageFileValidation struct {
	IsValid      bool   `json:"isValid"`
	ErrorMessage string `json:"errorMessage,omitempty"`
}

func NewImageFileValidation(err error) ImageFileValidation {
	if err == nil {
		return ImageFileValidation{IsValid: true}
	}
	return ImageFileValidation{
		IsValid:      false,
		ErrorMessage: err.Error(),
	}
}

// NewListImage creates a ListImage with validation.
//
// Firebase Storage 前提:
//   - url は Firebase Storage getDownloadURL() の値
//   - objectPath は Firebase Storage object path
//     例: lists/{listId}/images/{imageId}/{fileName}
func NewListImage(
	id string,
	listID string,
	u string,
	objectPath string,
	fileName string,
	contentType string,
	size int64,
	displayOrder int,
	createdAt time.Time,
	createdBy string,
) (ListImage, error) {
	li := ListImage{
		ID:           strings.TrimSpace(id),
		ListID:       strings.TrimSpace(listID),
		URL:          strings.TrimSpace(u),
		ObjectPath:   normalizeObjectPath(objectPath),
		FileName:     normalizeImageFileName(fileName),
		ContentType:  strings.TrimSpace(contentType),
		Size:         size,
		DisplayOrder: displayOrder,
		CreatedAt:    createdAt.UTC(),
		CreatedBy:    strings.TrimSpace(createdBy),
	}

	if li.ContentType == "" {
		li.ContentType = inferContentTypeFromFileName(li.FileName)
	}

	if err := li.Validate(); err != nil {
		return ListImage{}, err
	}

	return li, nil
}

// NewListImageWithCanonicalPath builds objectPath as:
//
//	lists/{listId}/images/{imageId}/{fileName}
func NewListImageWithCanonicalPath(
	id string,
	listID string,
	downloadURL string,
	fileName string,
	contentType string,
	size int64,
	displayOrder int,
	createdAt time.Time,
	createdBy string,
) (ListImage, error) {
	fn := normalizeImageFileName(fileName)

	return NewListImage(
		id,
		listID,
		downloadURL,
		CanonicalListImageObjectPath(listID, id, fn),
		fn,
		contentType,
		size,
		displayOrder,
		createdAt,
		createdBy,
	)
}

func (li *ListImage) UpdateURL(u string) error {
	u = strings.TrimSpace(u)
	if err := validateURL(u); err != nil {
		return err
	}

	li.URL = u
	return nil
}

func (li *ListImage) UpdateObjectPath(objectPath string) error {
	obj := normalizeObjectPath(objectPath)
	if err := validateObjectPath(obj); err != nil {
		return err
	}

	li.ObjectPath = obj
	return nil
}

func (li *ListImage) UpdateFileName(name string) error {
	fn := normalizeImageFileName(name)
	if !isAllowedImageExtension(fn) {
		return ErrInvalidListImageFileName
	}

	li.FileName = fn
	return nil
}

func (li *ListImage) UpdateContentType(contentType string) error {
	ct := strings.ToLower(strings.TrimSpace(contentType))
	if !isSupportedImageMIME(ct) {
		return ErrInvalidListImageContentType
	}

	li.ContentType = ct
	return nil
}

func (li *ListImage) UpdateSize(size int64) error {
	if size < 0 {
		return ErrInvalidListImageSize
	}

	if MaxListImageSize > 0 && size > MaxListImageSize {
		return ErrInvalidListImageSize
	}

	li.Size = size
	return nil
}

func (li *ListImage) SetDisplayOrder(order int) error {
	if order < 0 {
		return ErrInvalidListImageDisplayOrder
	}

	li.DisplayOrder = order
	return nil
}

func (li *ListImage) Touch(now time.Time, actor string) {
	if now.IsZero() {
		now = time.Now().UTC()
	}

	t := now.UTC()
	li.UpdatedAt = t
	li.UpdatedBy = strings.TrimSpace(actor)
}

// Validate validates a ListImage.
func (li ListImage) Validate() error {
	if li.ID == "" {
		return ErrInvalidListImageID
	}

	if li.ListID == "" {
		return ErrInvalidListImageListID
	}

	if err := validateURL(li.URL); err != nil {
		return err
	}

	if err := validateObjectPath(li.ObjectPath); err != nil {
		return err
	}

	if li.FileName == "" || !isAllowedImageExtension(li.FileName) {
		return ErrInvalidListImageFileName
	}

	if li.ContentType != "" && !isSupportedImageMIME(li.ContentType) {
		return ErrInvalidListImageContentType
	}

	if li.Size < 0 {
		return ErrInvalidListImageSize
	}

	if MaxListImageSize > 0 && li.Size > MaxListImageSize {
		return ErrInvalidListImageSize
	}

	if li.DisplayOrder < 0 {
		return ErrInvalidListImageDisplayOrder
	}

	if li.CreatedAt.IsZero() {
		return ErrInvalidListImageCreatedAt
	}

	if li.CreatedBy == "" {
		return ErrInvalidListImageCreatedBy
	}

	return nil
}

// ValidateDataURL validates a data URL: data:<mime>;base64,<payload>.
func ValidateDataURL(
	data string,
	maxBytes int,
	supported map[string]struct{},
) (mime string, payload []byte, err error) {
	if !strings.HasPrefix(data, "data:") {
		return "", nil, errors.New("invalid data URL: missing 'data:' prefix")
	}

	parts := strings.SplitN(data, ",", 2)
	if len(parts) != 2 {
		return "", nil, errors.New("invalid data URL: missing payload")
	}

	meta := parts[0]
	raw := parts[1]

	if !strings.Contains(meta, ";base64") {
		return "", nil, errors.New("invalid data URL: not base64 encoded")
	}

	mime = strings.TrimPrefix(strings.SplitN(meta, ";", 2)[0], "data:")
	if mime == "" {
		return "", nil, errors.New("invalid data URL: missing mime type")
	}

	if supported == nil {
		supported = SupportedImageMIMEs
	}

	if _, ok := supported[mime]; !ok {
		return "", nil, fmt.Errorf("unsupported content type: %s", mime)
	}

	decoded, decErr := base64.StdEncoding.DecodeString(raw)
	if decErr != nil {
		return "", nil, fmt.Errorf("invalid base64 payload: %w", decErr)
	}

	if len(decoded) == 0 {
		return "", nil, errors.New("empty image payload")
	}

	if maxBytes <= 0 {
		maxBytes = int(DefaultMaxImageSizeBytes)
	}

	if len(decoded) > maxBytes {
		return "", nil, fmt.Errorf("file too large: %d bytes (max %d)", len(decoded), maxBytes)
	}

	return mime, decoded, nil
}
